### DISCUSSION TITLE
write data back to anvil data table

### DISCUSSION THREAD

**Discussion by jr-d3-data** [anvilistas/tabulator#83] [Q&A]
**What I’m trying to do:**
writing data back from tabulator table to anvil data table:

1. having an editable cell, that, once filled, should write back to the anvil data table 
2. update of the tabulator table after the edit (i filter with the left panel on months, but switching back to the edited month, the inserted data is gone)

**What I’ve tried and what’s not working:**
tried to use the tabulator and anvil functions to write back the data persistantly

**Code Sample:**

```
 def tabulator_cell_edited(self, cell, **event_args):
    """This method is called when a cell is edited"""
    #self.tabulator.set_data()
    self.tabulator.update_data([{"id":cell.get_row().get_index(), cell.get_field():cell.get_value()}])
    
    curr_slot=cell.get_row().get_cell("timeslot").get_initial_value()
    
    curr_row = app_tables.plan.get(date=cell.get_row().get_cell("date"),slot=curr_slot)
    
    field=cell.get_field()
    # Set  new value
    curr_row.update(field=cell.get_value())

```
 
Clone of the app: https://anvil.works/build#clone:FUZMCJAXR3VC3FCL=XTER2ZAWCJF2BT6XJ4Y7EBU4

**Comment by s-cork** [comment]
The quick solution:
```python
    def tabulator_cell_edited(self, cell, **event_args):
        """This method is called when a cell is edited"""
        field = cell.get_field()
        row = cell.get_table_row()
        row[field] = cell.get_value()
```

**Why this works:**
When a cell is edited the internal data is up to date
So you never have to call `update_data`
you just need to update the table row to reflect the internal changes

**Why your code breaks:**
Thanks for the clone - very helpful

There's something weird going on with the line
`self.tabulator.update_data(...)`
Tabulator is swallowing this line - this looks like a bug in the Javascript Tabulator (I've reported it)
(They are never resolving a promise)

The problem is that you are not defining your index column.
It's best practice when using Tabulator to define your own index column.
Otherwise (when using the `app_table` option) Anvil Tabulator will create a private index field within the data.
So index field was actually `"_$id"`
i've now changed this so the index is whatever you're default id option was (in your case `"id"`)


But like I said - there's no need to update the data here since it's up to date

**Reply by jr-d3-data** [reply]
thanks @s-cork  for your input 🙌 yet, i am still having trouble...
to show it (as i am having a hard time writing it down), i made a little loom:
https://www.loom.com/share/2c9e646627604d589d9e8622b23df680
i hope this makes my two issues clear(er)... 
if not, let me know! :)

**Reply by s-cork** [reply]
Unfortunately I'm not able to open the link - it times out for me

**Reply by s-cork** [reply]
But I see the issue - I didn't address point 2
That looks like stale cache from my part
I'll get that fixed and update you

**Reply by s-cork** [reply]
Hopefully this now works as expected
I've also updated the docs for how to clear the cache manually.

**Reply by s-cork** [reply]
@jr-d3-data did you check if it's now working?
